backend/expandToppingsCategories.ts

```diff
@@ -0,0 +1,30 @@
+import { getCategoriesCollection } from './db'
+import { Topping, ToppingWithCategoryId } from './topping'
+
+export default async function expandToppingsCategories(
+  toppings: ToppingWithCategoryId[]
+): Promise<Topping[]>
+{
+  const categoriesCollection = await getCategoriesCollection()
+  const categoriesCache: Record<number, string> = {}
+
+  async function getCategoryName(categoryId: number) {
+     if (!(categoryId in categoriesCache)) {
+      const category = await categoriesCollection.findOne({id: categoryId})
+      if (category === null) {
+        throw new Error('Topping category doesn\'t exist')
+      }
+      categoriesCache[categoryId] = category.name
+    }
+    return categoriesCache[categoryId]
+  }
+
+  return Promise.all(
+    toppings.map(
+      async topping => ({
+        ...topping,
```

Comment:
what is the ... for?

Reply:
it unpacks the contents of `topping` into the surrounding curly brackets

the whole bit inside the curly brackets makes a new topping with everything the same as the old topping except with `category` set to a string with the name rather than the id number

Reply:
clever 🙃